admin/class-admin-apple-sections.php

```diff
@@ -0,0 +1,364 @@
+	 */
+	private $settings;
+
+	/**
+	 * Valid actions handled by this class and their callback functions.
+	 *
+	 * @var array
+	 * @access private
+	 */
+	private $valid_actions;
+
+	/**
+	 * Returns a taxonomy object representing the taxonomy to be mapped to sections.
+	 *
+	 * @access public
+	 * @return WP_Taxonomy|false A WP_Taxonomy object on success; false on failure.
+	 */
+	public static function get_mapping_taxonomy() {
+
+		/**
+		 * Allows for modification of the taxonomy used for section mapping.
+		 *
+		 * @since 1.2.2
+		 *
+		 * @param string $taxonomy The taxonomy slug to be filtered.
+		 */
+		$taxonomy = apply_filters( 'apple_news_section_taxonomy', 'category' );
+
+		return get_taxonomy( $taxonomy );
+	}
+
+	/**
+	 * Given a post ID, returns an array of section URLs based on applied taxonomy.
+	 *
+	 * Supports overrides for manual section selection and fallback to postmeta
+	 * when no mappings are set.
+	 *
+	 * @param int $post_id The ID of the post to query.
+	 *
+	 * @access public
+	 * @return array An array of section URLs for the post.
+	 */
+	public static function get_sections_for_post( $post_id ) {
+
+		// Determine if sections should be loaded from postmeta.
+		$mappings = get_option( self::TAXONOMY_MAPPING_KEY );
+		$override = get_post_meta( $post_id, 'apple_news_section_override', true );
+		if ( empty( $mappings ) || ! empty( $override ) ) {
+			return get_post_meta( $post_id, 'apple_news_sections', true );
+		}
+
+		// Try to get sections.
+		$admin_settings = new Admin_Apple_Settings;
+		$section_api = new Section( $admin_settings->fetch_settings() );
+		$sections_raw = $section_api->get_sections();
+		if ( empty( $sections_raw ) || ! is_array( $sections_raw ) ) {
+			wp_die( __( 'Unable to fetch a list of sections.', 'apple-news' ) );
+		}
+
+		// Convert sections returned from the API into a key/value pair of id/URL.
+		$sections = array();
+		foreach ( $sections_raw as $section ) {
+			if ( ! empty( $section->id ) && ! empty( $section->links->self ) ) {
+				$sections[ $section->id ] = $section->links->self;
+			}
+		}
+
+		// Try to get configured taxonomy.
+		$taxonomy = self::get_mapping_taxonomy();
+		if ( empty( $taxonomy ) || is_wp_error( $taxonomy ) ) {
+			wp_die( __( 'Unable to get a valid mapping taxonomy.', 'apple-news' ) );
+		}
+
+		// Try to get terms for the post.
+		$terms = get_the_terms( $post_id, $taxonomy->name );
+		if ( empty( $terms ) || is_wp_error( $terms ) ) {
+			return array();
+		}
+
+		// Loop through the mappings to determine sections.
+		$post_sections = array();
+		$term_ids = wp_list_pluck( $terms, 'term_id' );
+		foreach ( $mappings as $section_id => $section_term_ids ) {
+			foreach ( $section_term_ids as $section_term_id ) {
+				if ( in_array( $section_term_id, $term_ids, true ) ) {
+					$post_sections[] = $sections[ $section_id ];
+				}
+			}
+		}
+
+		// Eliminate duplicates.
+		$post_sections = array_unique( $post_sections );
+
+		return $post_sections;
```

Comment:
Are we storing these somewhere so we don't have to execute this logic from line 99 onward every time?